### ISSUE TITLE
Time series data with a combination of continuous and categorical features

### ISSUE THREAD

**Issue by tkharrat** [timeseriesAI/tsai#231]
Hi,

I have a dataset with a sequence of continuous and categorical variables. The aim is to classify these sequences. I searched the tutorials notebooks but I can't find an example where this scenario is discussed. 

In my opinion, a strategy similar to what has been done in fastai tabular_learner can be considered. The the categorical tesnors are first passed through an embedding and then concatenated with the continuous features. Next, the resulting tensor can be passed through the rest of the network.

If this scenario has been considered already, I would be grateful if you could point me to the associated notebook. If not, any guidance on how it could be implemented would be very useful. I would be keen to submit a PR dealing with this as I am sure some users may have similar datasets.

Thanks

**Comment by Wabinab** [comment]
Oh right @oguiza the problem you describe is different from what one encountered. Sorry for the misinterpretation. My problem/requirement is similar to @tkharrat. The time series data is of shape `(seq_len, num_features)` so that at each time steps, there are `num_features` describing that particular time step. So one would have X of size `(batch_size, seq_len, num_features)` and y (target) of shape `(batch_size, seq_len)` (at least for one's problem, but it might well be other shape depending on what regression/classification problem one is trying to deal with, so y's shape isn't too rigid). 

So the shape of X would be as below (or its transpose)

| | time step 1 | time step 2 | time step 3 | ... |
| --- | --- | --- | --- | --- |
| feature 1 | ... | ... | ... | ... |
| feature 2 | ... | ... | ... | ... |
| feature 3 | ... | ... | ... | ... |
| ... | ... | ... | ... | ... |

One haven't thought of any good idea of how to solve the categorical problems. Though NLP uses embeddings one haven't encountered any problems NLP with multiple features except for a sentence/paragraph passed in as input, so that's really is one single feature.

**Comment by Wabinab** [comment]
Hi @oguiza thanks for the update. As one promises one made the notebook here: https://colab.research.google.com/drive/1rwA8lzCoz_PpIb9qceADSM8OUuwxIwBX?usp=sharing you could check it out how one does it for the moment. And @oguiza no one doesn't find any dataset that kinda resembles this if we cannot use competition dataset then perhaps we might need to find deeper or more easier, just make one ourselves as you mentioned. 

Thanks again for your work @oguiza

**Comment by tkharrat** [comment]
Hi @oguiza 

The data provided is a list of tuples as described above. One thing to note though is that the sequence of play (the first element in the tuple) are of different lengths (i.e number of row in the dataFrame). That's why I mentioned padding in my previous post (which will happen in the `before_batch` step). Therefore, I am not sure we can rely on the combination of `preprocess_df` and `df2Xy` whcih makes the assumption of the same number of time step in every example. Besides, `preprocess_df` does not handle the `splits` and process the full dataFrame which may create future information leakage.

 I think for this task, it may be easier to write a custom `tfmdLists` that has a proper show method and create a batch with 3 tensors `(cat, cont, target)`. I think if we store `cat` and `cont` in the same tensor, we will have to convert `cat` to float (from integer), and then re-split into 2 tensors because the `cat` will have to go through an embedding first.  

However, if we present the data in a 3 tensors format, I think we could use any tsai learner with a small amendment.

- Inherit from the desired Learner class 
- Slightly amend the `forward()` method: It will have to take the `cat` tensor, pass it to embedding, concatenate it with `cont` and then call the mother class `forward()` method.

The main drawback of this approach is that it may not be generalizable to any sequence data with cat and cont features (user will have to prepare his own `tfmdLists`) but it could be used as an example of how it can be done. 

I am happy to make an attempt and share the notebook. Maybe @oguiza it can help you generalise it even further.

**Comment by oguiza** [comment]
Hi @tkharrat,
I've been looking at the dataset and notebook you have created. I find it too complicated for a notebook tutorial. The goal of tutorials is to demonstrate how you can easily apply a model or method to your own dataset.
So I'd like to propose something that might still be useful: 
1. Create a dataframe using one of the UCR multivariate datasets.
2. Modify the scale. This will allow us to show demonstrate how to scale the data using different techniques (standardization, normalization, etc) while preventing leakage.
3. Add categorical data. This will allow us to demonstrate the use of categorify. And later demonstrate the use of embeddings.
4. We can also modify the time series length so that they are different. This will allow the use of padding and truncating.
5. We can also add delete some data to simulate nan values. 
This will require df2Xy to be modified to allow the padding/ truncating.

Once the data is ready, we can create the dataloaders. I believe minor changes will be necessary. In my opinion, we'll be able to use a single dataloader for both categorical and continuous data. 

This approach will require a change in the models (which I'm willing to make anyways) to adapt them to take and separate categorical from continuous data (I've tested it with an attribute cat_pos for categorical positions and cont_pos for continuous positions and it works well). I've also built a MultiEmbedding layer that needs to be added to the models (this will allow the creation of multiple, concatenate embeddings).

I believe all this is something that can be done fairly quickly. We can create a simple tutorial with this approach. You can then test this approach with the large dataset you proposed. And if you find any issues, we may need to make adjustments. 

As I said, I am willing to make this available in tsai as I think it is a common scenario. Please let me know what you think and if you are still willing to collaborate on this.

**Comment by tkharrat** [comment]
Hi @oguiza,

The plan above sounds good. I just want to add few comments:

1) if you want to load your data in a unique tensor, and use an attribute cat_pos for categorical positions and cont_pos for continuous, that's fine but you will have to convert int to float in the cat_pos locations. No big deal obviously but it does not seem to be the fastai philosophy.

2) on the other hand, if we follow the fastai tabular approach and return our data in the form of tuples of length 3 (cats, conts, target), then I think modifying the models becomes super easy. We just need to create a head that applies the embedding to cats and concatenates the results with conts. The obtained tensor can be processed by the forward() method of any model already defined in tsai.

I made an attempt of implementing 2) [here](https://github.com/tkharrat/footSeq/blob/master/01_data_structure.ipynb). It does not seem too bad.

I am happy with whichever path you want to take. Just let me know how to proceed and how I can help and I am more than keen to contribute.

**Comment by oguiza** [comment]
Hi @tkharrat,

Thanks for sharing your code. I think it looks very good. 

Personally, I think both options are kind of similar. To make a model split a tensor between categorical and continuous variables, or take a tuple is easy. I can adapt the tsai models for either option (or both).
The most difficult part is to create a pipeline that transforms a dataframe with categorical and continuous data into batches. I can see 2 options: 

1. Transform the df into a numpy array using any type of sklearn preprocessors. Once we have the array we pass it to a tsai dataloader.
2. Create a dataloader using fastai's tabular data approach. 

One of the reasons I created tsai's TSDataLoaders is that fastai's way of handling arrays was very slow. It processes each sample individually and applies the transforms on the fly. tsai preprocesses all samples during the dataset creation and then manages all samples in the batch at the same time. That's why batch creation is about 100 times faster using tsai compared to the vanilla, DataBlock process. I don't know, however, how fast the tabular data process is. But it's something to bear in mind. 
Here's what I propose: 
1. Let's use the dataset I shared in the [gist](https://gist.github.com/oguiza/c7b2f0262c7eb2f0a89e8234bd0903ec). I think it has all the required features. 
2. Let's find o several methods to create a dataloader. And let's measure the dataloader performance. One of the methods could be the one you have used in your notebook. But ideally, we should avoid customizing the solution as much as we can. Although there will be some transforms that will be required (as the padding transform).
3. When we have that, I'll make the required updates in tsai's model to allow them to take batches of categorical & continuous data (as a single tensor or a tuple of tensors). 

Once we get this to work well, you'll need to create a custom Tensor subclass with a show method to display your data. 

Please, let me know if you are ok with this approach. 

I'll start working on #2 now.

**Comment by tkharrat** [comment]
Hi @oguiza,

I think this plan makes sense. Actually, you already solved point n3 with [MultiEmbedding](https://github.com/timeseriesAI/tsai/blob/main/tsai/models/layers.py#L1105). With minimum coding, I think we can extend any tsai model to fit the current set up.

I will carry on experienting on my side and report back.

**Comment by oguiza** [comment]
Hi @tkharrat,

I've made some progress on this task. Here's the approach I'm working on. I think it'd be good to split data preparation into separate components:
    
   -  initial data preparation: this encompasses all tasks at a dataframe level. These transforms are usually implemented once and don't need to be reversible. In fastai, these types of tasks are handled with separate functions (like make_date, add_datepart, add_elapsed_times, df_shrink, etc). But there are many others like feature engineering, missing data encoder, etc. All this will happen before a tsai dataset is built. This will give the user maximum flexibility. 
   - transform the dataframe into X and y arrays (using SlidingWindowPanel). The output of this could be a numpy array in memory or on disk (useful for large datasets that don't fit in memory). And sequences may have different lengths.
   - once the arrays are converted into tensors using a TSDataset, apply batch transforms. There are several functions already available in tsai that can be used to scale the data like Standardize, Normalize, etc that can be used. 
 
TODO:

- initial data preparation: there are many functions in sklearn that can be used to build a pipeline. But other functions may be required. I've already prepared a few that are not available in sklearn.

- df to Xy: SlidingWindowPanel may be used to create the X and y arrays from the data. However, I've identified a couple of areas of improvement that I'll implement soon. I'd like to know your opinion on this. Let's analyze different scenarios with a window length of 30 in all of them: 
   - a sequence is 100 steps long. How should we apply the sliding window? I'd say create 3 windows with positions: 11-40, 41-70, and 71-100, and discard 0-10. This is not the way it works now (1-30, 31-60, and 61-90, and discard 91-100).
   - s sequence is 20 steps long. Option 1: discard the sample since it's too short. Option 2: use it padding at the end. Option 3: use it padding at the beginning. I think option 3 makes more sense.
   - a sequence has 2 steps. Should a minimum number of steps be set? I'd say it makes sense to set a minimum percentage.

- The models will be updated to accept both categorical and continuous data.

I'd love to have your view on all this.

**Comment by vrodriguezf** [comment]
> I think option 3 makes more sense

I wonder, why does it make more sense to pad at the beginning than at the end?

> a sequence has 2 steps. Should a minimum number of steps be set? I'd say it makes sense to set a minimum percentage.

Absolutely!

**Comment by oguiza** [comment]
Thanks for sharing your thoughts @vrodriguezf!
My rationale is that is we are preparing a dataset with a window length of 30 but we only have 20 timesteps for a given instance, I would think we are missing some history (at the beginning), and only have the 20 last steps (you may want to predict the next one for example in forecasting). It's true that it may not matter much in a classification or regression task. 
What do think?

**Comment by vrodriguezf** [comment]
Mmm ok I see the point, when you have to choose what to keep, you give priority to the future, because at the end, it's what matters the most generally in the time series world.

Anyway, for the first case you propose (100 time steps), there is no need to discard, just padding, right?

**Comment by oguiza** [comment]
>Anyway, for the first case you propose (100 time steps), there is no need to discard, just padding, right?

It depends on whether a minimum unpadded length is established or not. Bear in mind that the "remainder" might just be 1 time step. And you don't want to use t 1 step padded up to 30.  I proposed to set a limit (which you seemed to agree to). So in the end the solution might be to maintain the pad_remainder arg but add a min_remainder_size (or similar). And change the logic to start from the end so that the remainder is always the oldest part of the sequence.

**Comment by vrodriguezf** [comment]
oh yes sure, it's the same situation in both cases :) looks coherent to me!!

**Comment by oguiza** [comment]
cc: @tkharrat, @vrodriguezf 

I'd like to update the status of this enhancement request. I believe all the different components required to perform this type of task are in place now. I've started to use it on some proprietary datasets and it seems to work well. 
I've made the following additions to ```tsai```:

- Updated SlidingWindow and SlidingWindowPanel so they can now pad sequences starting from either end.
- Added TSCategoricalEncoder that allows the creation of categories for 1 or more columns in a pandas dataframe. It has a state so you can fit on the train set, and then transform train, validation, and test data.
- Added MultiEmbedding layer that applies an embedding layer to each variable passed as categorical, concatenating the output to continuous features. 
- Added cat_pos (categorical variable position), n_embeds (number of embeddings) and emb_szs (embedding sizes) to LSTMPlus and TSiTPlus. This may be extended to other models if necessary.

The main task that is remaining IMO is a tutorial to demonstrate how this functionality can be used. I will try to do this over the coming weeks.

**Comment by vrodriguezf** [comment]
Wow, amazing work Ignacio!!! This is a huge addition to the library.

> It has a state so you can fit on the train set, and then transform train, validation, and test data

what does this mean? Does this have to do with the situation where some category's of a variable may not appear in training, but in test time?

**Comment by oguiza** [comment]
> Wow, amazing work Ignacio!!! This is a huge addition to the library.

I think so. I've started to use categorical embeddings on some datasets, and it works pretty well.

> > It has a state so you can fit on the train set, and then transform train, validation, and test data
>
> what does this mean? Does this have to do with the situation where some category's of a variable may not appear in training, but in test time?

Exactly. You need to ensure you calculate categorical embeddings in the same way you calculated them for training. It's not just that they may be missing. They might also be in a different order.

**Comment by strakehyr** [comment]
> cc: @tkharrat, @vrodriguezf
> 
> I'd like to update the status of this enhancement request. I believe all the different components required to perform this type of task are in place now. I've started to use it on some proprietary datasets and it seems to work well. I've made the following additions to `tsai`:
> 
> * Updated SlidingWindow and SlidingWindowPanel so they can now pad sequences starting from either end.
> * Added TSCategoricalEncoder that allows the creation of categories for 1 or more columns in a pandas dataframe. It has a state so you can fit on the train set, and then transform train, validation, and test data.
> * Added MultiEmbedding layer that applies an embedding layer to each variable passed as categorical, concatenating the output to continuous features.
> * Added cat_pos (categorical variable position), n_embeds (number of embeddings) and emb_szs (embedding sizes) to LSTMPlus and TSiTPlus. This may be extended to other models if necessary.
> 
> The main task that is remaining IMO is a tutorial to demonstrate how this functionality can be used. I will try to do this over the coming weeks.

Hi @oguiza,
Is there a chance you can post here a short chunk of code showing the TSCategorical functionality, side by side with a TSRegression, for instance? I understand a fully fleshed-out tutorial would be better, but a few guiding steps could be very helpful.

**Comment by oguiza** [comment]
Closing this issue due to lack of activity and progress. If necessary please, create a new one.
